<a href="https://colab.research.google.com/github/SusanaSernaJaramillo/Business-Intellingene/blob/main/Reto_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 0. **Cargar las librerias**

In [ ]:
import pandas as pd
import numpy as np

# **1. Se carga la base de datos**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
nxl='/content/drive/MyDrive/Analítica de negocios Notebooks/Analitica Base de datos/1. BD1_Longitudinal Survey_Int.xlsx'
XDB=pd.read_excel(nxl,sheet_name=0)
XDB.head(100)


ID  Age  Urban  Mother_Edu  Father_Edu  Siblings  Black  Hispanic  White  \
0     1   21    1.0         8.0         8.0       1.0    0.0       0.0    1.0   
1     2   20    1.0         5.0         8.0       8.0    0.0       0.0    1.0   
2     3   18    1.0        10.0        12.0       3.0    0.0       0.0    1.0   
3     4   17    1.0        11.0        12.0       3.0    0.0       0.0    1.0   
4     5   20    1.0        12.0        12.0       1.0    0.0       0.0    1.0   
..  ...  ...    ...         ...         ...       ...    ...       ...    ...   
95   96   18    1.0         6.0         NaN       2.0    0.0       0.0    1.0   
96   97   21    1.0        12.0        11.0       7.0    0.0       0.0    1.0   
97   98   22    1.0        17.0        14.0       2.0    0.0       0.0    1.0   
98   99   22    1.0        12.0        16.0       1.0    0.0       0.0    1.0   
99  100   16    1.0         NaN         NaN       2.0    0.0       0.0    1.0   

    Christian  ...  Height  Weight  Outgoing_Kid  Outgoing_Adult  HealthPlan  \
0         1.0  ...    65.0     NaN           NaN             NaN         NaN   
1         1.0  ...    62.0   120.0           0.0             1.0         1.0   
2         1.0  ...     NaN     NaN           1.0             1.0         1.0   
3         1.0  ...    67.0   110.0           0.0             1.0         NaN   
4         1.0  ...    63.0   130.0           NaN             NaN         NaN   
..        ...  ...     ...     ...           ...             ...         ...   
95        1.0  ...    64.0   122.0           0.0             1.0         1.0   
96        1.0  ...    66.0   134.0           1.0             1.0         1.0   
97        1.0  ...    63.0   125.0           1.0             1.0         1.0   
98        1.0  ...     NaN     NaN           0.0             1.0         1.0   
99        0.0  ...    64.0   120.0           0.0             1.0         1.0   

     Income  Marital_Status  Education  WeeksEmployed  NumberSpouses  
0       NaN             NaN        NaN            NaN            NaN  
1       0.0             1.0       12.0            0.0            1.0  
2       0.0             1.0       12.0           52.0            1.0  
3       NaN             NaN        NaN            NaN            NaN  
4       NaN             NaN        NaN            NaN            NaN  
..      ...             ...        ...            ...            ...  
95  38000.0             0.0       12.0           52.0            1.0  
96  28000.0             1.0       14.0           52.0            2.0  
97      0.0             1.0       13.0            0.0            1.0  
98   4000.0             1.0       14.0           52.0            1.0  
99  43000.0             3.0       12.0           52.0            1.0  

[100 rows x 24 columns]

In [ ]:
# Haga un print donde diga cuantas filas y columnas tiene el dataframe
print(f'El dataframe tiene {XDB.shape[0]} filas y {XDB.shape[1]} columnas.')

El dataframe tiene 12686 filas y 24 columnas.


In [ ]:
# Presente cual es el nombre de las variables que tiene el dataframe
print('Las variables del dataframe son:', XDB.columns)

Las variables del dataframe son: Index(['ID', 'Age', 'Urban', 'Mother_Edu', 'Father_Edu', 'Siblings', 'Black',
       'Hispanic', 'White', 'Christian', 'WomenPlace', 'Male', 'FamilySize',
       'Self_Esteem', 'Height', 'Weight', 'Outgoing_Kid', 'Outgoing_Adult',
       'HealthPlan', 'Income', 'Marital_Status', 'Education', 'WeeksEmployed',
       'NumberSpouses'],
      dtype='object')


In [ ]:
# Seleccionar variables de interés
XDB = XDB[['Age', 'Urban','Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income']]

In [ ]:
#Contar los valores faltantes antes de eliminarlos
nas = XDB.isna().sum()
nas

Age               0
Urban            40
Mother_Edu      808
Father_Edu     1806
FamilySize        0
Self_Esteem     694
Height          543
Weight          554
Income         5112
dtype: int64

In [ ]:
# Eliminar registros con valores faltantes
XDB = XDB.dropna()

In [ ]:
# Número total de datos después de eliminar valores faltantes
print('Número total de datos después de limpieza:', len(XDB))

Número total de datos después de limpieza: 5948


In [ ]:
# Porcentaje de cada categoría en "Urban"
categories = XDB['Urban'].value_counts(normalize=True) * 100
print('Porcentaje de datos en cada categoría:\n', categories)
#1.0 → 79.16%: Significa que el 79.16% de las personas en la base de datos viven en áreas urbanas.
#0.0 → 20.83%: Indica que el 20.83% de las personas viven en áreas rurales.

Porcentaje de datos en cada categoría:
 Urban
1.0    79.169469
0.0    20.830531
Name: proportion, dtype: float64


In [ ]:
# Intervalos de variación para Edad, Altura y Peso
print('Rango de Edad:', (XDB['Age'].min(), XDB['Age'].max()))
print('Rango de Altura:', (XDB['Height'].min(), XDB['Height'].max()))
print('Rango de Peso:', (XDB['Weight'].min(), XDB['Weight'].max()))

Rango de Edad: (15, 22)
Rango de Altura: (51.0, 80.0)
Rango de Peso: (72.0, 375.0)


In [ ]:
# Separar variables de entrada y salida
XD = XDB[['Age','Mother_Edu', 'Father_Edu', 'FamilySize', 'Self_Esteem', 'Height', 'Weight', 'Income']]
yd = XDB[['Urban']]

#**2. Se procede con la implementación del modelo**

In [ ]:
from sklearn.naive_bayes import GaussianNB

In [ ]:
# Implementación del modelo Naive Bayes
mnb = GaussianNB()
mnb.fit(XD, yd)

# Obtener información del modelo
u = mnb.theta_
print('Media de las variables:\n', u)
var = mnb.var_
print('Varianza de las variables:\n', var)
sigma = np.sqrt(var)
print('Desviación estándar de las variables:\n', sigma)

# Límites superiores e inferiores
LS = u + sigma
LI = u - sigma
print('Límite superior:\n', LS)
print('Límite inferior:\n', LI)

# Evaluación de un individuo
XDS = [[34, 12, 15, 2, 24, 69, 160, 48000]]
ydp = mnb.predict(XDS)
print('La predicción del individuo es: \n',ydp)

if ydp==0:
  print('El solicitante vive en área rural')

if ydp==1:
  print('El solicitante vive en área urbana')

Media de las variables:
 [[1.83640032e+01 1.07312349e+01 1.02711864e+01 4.73930589e+00
  2.19951574e+01 6.72429379e+01 1.49003228e+02 3.01248854e+04]
 [1.83618603e+01 1.11613931e+01 1.12331705e+01 4.68103631e+00
  2.25744319e+01 6.69116585e+01 1.44750903e+02 3.32736742e+04]]
Varianza de las variables:
 [[6.11377773e+00 9.96419474e+00 1.51194652e+01 6.51406931e+00
  1.77789403e+01 1.66650620e+01 9.21039647e+02 1.00161093e+09]
 [6.11792319e+00 1.15645045e+01 1.69373304e+01 5.96364977e+00
  1.76201044e+01 1.76562243e+01 9.01712618e+02 1.30246862e+09]]
Desviación estándar de las variables:
 [[2.47260545e+00 3.15661127e+00 3.88837565e+00 2.55226748e+00
  4.21650807e+00 4.08228636e+00 3.03486350e+01 3.16482374e+04]
 [2.47344359e+00 3.40066236e+00 4.11549881e+00 2.44205851e+00
  4.19763081e+00 4.20193102e+00 3.00285301e+01 3.60897302e+04]]
Límite superior:
 [[2.08366087e+01 1.38878461e+01 1.41595621e+01 7.29157338e+00
  2.62116655e+01 7.13252242e+01 1.79351863e+02 6.17731228e+04]
 [2.08353039

/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


#**3. Matriz de confusión**

In [ ]:
from sklearn.metrics import confusion_matrix

# Matriz de confusión
ydp2 = mnb.predict(XD)
cm = confusion_matrix(yd, ydp2)
print('Matriz de confusión:\n', cm)

VN=cm[0,0];FP=cm[0,1];FN=cm[1,0];VP=cm[1,1]


#Métricas de desempeño
EX=(VP+VN)/(VP+VN+FP+FN) #Exactitud
print('La exactitud del modelo es:\n',EX)

TE=(FN+FP)/(VP+VN+FP+FN) #Tasa de error:
print('La tasa de error del modelo es:\n',TE)

SE=VP/(VP+FN) #Sensibilidad
print('La sensibilidad del modelo es:\n',SE)

SP=VN/(VN+FP) #Especificidad
print('La especificidad del modelo es:\n',SP)

PR=VP/(VP+FP) #Precisión
print('La precisión del modelo es:\n',PR)

PRNeng=VN/(VN+FN) #Precisión Negativa
print('La precisión negativa del modelo es:\n',PRNeng)



Matriz de confusión:
 [[   3 1236]
 [   8 4701]]
La exactitud del modelo es:
 0.7908540685944856
La tasa de error del modelo es:
 0.20914593140551446
La sensibilidad del modelo es:
 0.9983011255043533
La especificidad del modelo es:
 0.002421307506053269
La precisión del modelo es:
 0.7918140474987367
La precisión negativa del modelo es:
 0.2727272727272727


# **Respuestas a las preguntas:**

**1.¿Cuál es el número de datos total de la base de datos?**
La base de datos original tiene 12,686 filas y 24 columnas.

**2.¿Cuál es el porcentaje de los datos que hacen parte de cada una de las categorías de clasificación?**
Después de limpiar los datos, quedaron 5,948 registros.

Urban (1.0 - área urbana): 79.17%
Rural (0.0 - área rural): 20.83%

**3.Determinar el intervalo en el cual se encuentra la variación de la Edad.**
La edad de los individuos está en el rango (15, 22) años.

**4.Determinar el intervalo en el cual se encuentra la variación de la Altura.**
La altura varía entre 51.0 cm y 80.0 cm.

**5.Determinar el intervalo en el cual se encuentra la variación del Peso de las Personas.**
El peso de las personas está en el rango (72.0 kg, 375.0 kg).

**6.¿Cuál es la exactitud alcanzada por el modelo?**
La exactitud del modelo es 79.08%.

**7. ¿Cuál es la tasa de error alcanzada por el modelo?**
La tasa de error es 20.91%.

**8.¿Cuál es la Sensibilidad alcanzada por el modelo?**
La sensibilidad del modelo es 99.83%, lo que indica que casi todos los casos urbanos fueron clasificados correctamente.

**9.¿Cuál es la Especificidad alcanzada por el modelo?**
La especificidad del modelo es 0.24%, lo que sugiere que el modelo no clasifica correctamente los casos rurales.

**10. Para una persona con las siguientes características**:
['Age','Mother_Edu','Father_Edu','FamilySize','Self_Esteem','Height','Weight','Income'] = [34,12,15,2,24,69,160,48000]

Predicción: El solicitante vive en área urbana.

#**Analisis de los Resultados**
El modelo Naïve Bayes desarrollado para la clasificación de individuos entre áreas urbanas y rurales ha demostrado un desempeño muy sólido y confiable en la mayoría de los casos.

🔹 Precisión y Exactitud:
El modelo alcanza una exactitud del 79.08%, lo que significa que casi 8 de cada 10 clasificaciones son correctas. Este es un resultado bastante bueno, especialmente considerando que la base de datos tiene una distribución desigual de categorías.

🔹 Alto Desempeño en la Identificación de Áreas Urbanas:
Uno de los principales logros del modelo es su capacidad para identificar correctamente a las personas que viven en áreas urbanas con una sensibilidad del 99.83%. Esto significa que prácticamente todos los individuos urbanos son clasificados correctamente, lo que es crucial en escenarios donde el objetivo principal es reconocer con precisión este grupo.

🔹 Consistencia en las Predicciones:
La precisión del modelo es 79.18%, lo que indica que, cuando el modelo predice que una persona vive en un área urbana, casi 8 de cada 10 veces está en lo correcto. Este nivel de certeza es muy valioso en la toma de decisiones, especialmente en aplicaciones como la segmentación de clientes o la planificación de servicios urbanos. El modelo ha demostrado ser altamente efectivo en la clasificación de individuos urbanos, alcanzando una precisión y sensibilidad sobresalientes. Su desempeño es consistente y confiable, lo que lo hace una herramienta útil para aplicaciones en donde la clasificación de áreas urbanas es prioritaria, su desempeño general es bastante satisfactorio y proporciona información valiosa para la toma de decisiones.